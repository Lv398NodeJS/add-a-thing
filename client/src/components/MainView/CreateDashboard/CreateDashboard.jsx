import React, { Component } from 'react';
import {
  Button,
} from 'react-bootstrap';
import ModalToCreateDashboard from './ModalToCreateDash/ModalToCreateDash';
import '../../../App.scss';

export default class CreateDashboard extends Component {
  constructor() {
    super();

    this.state = {
      showComponent: false,
      showLogin: false,
    };
  }

  toggleModal = () => {
    const { showComponent } = this.state;
    this.setState({
      showComponent: !showComponent,
    });
  };

  closeModal = () => {
    const { showLogin } = this.state;
    this.setState({
      showLogin: !showLogin,
    });
  };

  render() {
    const { showComponent } = this.state;
    return (
      <>
        <Button
          className="createDash"
          variant="primary"
          onClick={this.toggleModal}
        >
          Create new dashboard
        </Button>
        <ModalToCreateDashboard
          closeModal={this.toggleModal}
          show={showComponent}
          className="modal-to-create"
        />
      </>
    );
  }
}
