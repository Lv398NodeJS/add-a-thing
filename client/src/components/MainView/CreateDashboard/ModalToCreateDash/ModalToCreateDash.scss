.modal-form {
  width: 90%;
  margin-left: 5%;
  margin-right: 5%;
  margin-top: 10;
}
