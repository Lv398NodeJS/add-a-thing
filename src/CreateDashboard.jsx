import React, { Component } from 'react';
import './App.scss';
import { Button, Modal, Form } from 'react-bootstrap';

export default class CreateDashboard extends Component {
  constructor() {
    super();

    this.handleShow = this.handleShow.bind(this);
    this.handleClose = this.handleClose.bind(this);

    this.state = {
      show: false,
    };
  }

  handleClose() {
    this.setState({ show: false });
  }

  handleShow() {
    this.setState({ show: true });
  }

  render() {
    const styles = {
      width: '90%',
      marginLeft: '5%',
      marginRight: '5%',
    };
    const { show } = this.state;
    return (
      <content className="App">
         <Button variant="primary" onClick={this.handleShow}>
          Click to create new dashboard...
        </Button>

        <Modal
          size="lg"
          show={show} 
          onHide={this.handleClose}
        >
          <Modal.Header closeButton>
            <Modal.Title>Create Dashboard</Modal.Title>
          </Modal.Header>
          <Form.Control style={styles} type="text" placeholder="Enter the name for the new dashborad" />
          <br />
          <Modal.Body>Click 'Save Changes' to create a new dashboard</Modal.Body>
          <Modal.Footer>
            <Button variant="secondary" onClick={this.handleClose}>
              Close
            </Button>
            <Button variant="primary" onClick={this.handleClose}>
              Save Changes
            </Button>
          </Modal.Footer>
        </Modal>
      </content>
    );
  }
}
