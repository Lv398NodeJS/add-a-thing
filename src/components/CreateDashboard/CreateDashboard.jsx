import React, { Component } from 'react';
import '../../App.scss';
import { Button, Modal, Form } from 'react-bootstrap';

export default class CreateDashboard extends Component {
  constructor() {
    super();

    this.handleShow = this.handleShow.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.handleSave = this.handleSave.bind(this);

    this.state = {
      show: false,
      dashName: '',
      dashDescription: '',
    };
  }

  handleSaveButtonPush = () => {
    const { handler } = this.props;
    const { dashName, dashDescription } = this.state;
    handler({ dashName, dashDescription });
    this.setState({ show: false });
  }

  handleShow() {
    this.setState({ show: true });
  }

  handleSave(event) {
    const { name, value } = event.target;
    this.setState({
      [name]: value,
    });
  }

  handleClose() {
    this.setState({
      show: false,
    });
  }

  render() {
    const styles = {
      width: '90%',
      marginLeft: '5%',
      marginRight: '5%',
      marginTop: 10,
    };
    const { show, dashName, dashDescription } = this.state;
    return (
      <content className="App">
        <Button variant="primary" onClick={this.handleShow}>
          Click to create new dashboard...
        </Button>

        <Modal
          size="lg"
          show={show} 
          onHide={this.handleClose}
        >
          <Modal.Header closeButton>
            <Modal.Title>Create Dashboard</Modal.Title>
          </Modal.Header>
          <Form.Control
            style={styles}
            type="text"
            placeholder="Enter the name for the new dashboard"
            name="dashName"
            value={dashName}
            onChange={this.handleSave}
          />
          <Form.Group controlId="exampleForm.ControlTextarea1">
            <Form.Control
              style={styles}
              as="textarea"
              placeholder="Enter the description for the new dashboard"
              name="dashDescription"
              value={dashDescription}
              onChange={this.handleSave}
            />
          </Form.Group>
          <br />
          <Modal.Body>Click 'Save Changes' to create a new dashboard</Modal.Body>
          <Modal.Footer>
            <Button variant="secondary" onClick={this.handleClose}>
              Close
            </Button>
            <Button variant="primary" onClick={this.handleSaveButtonPush}>
              Save Changes
            </Button>
          </Modal.Footer>
        </Modal>
        {console.log(this.state)}
      </content>
    );
  }
}
