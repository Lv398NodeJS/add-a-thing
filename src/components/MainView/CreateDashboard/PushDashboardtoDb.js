import db from '../../../fire';

export default (dashName, dashDescription, addDashboard) => {
  const dashboardsRef = db.database().ref('dashboards');
  const dashboard = {
    name: dashName,
    description: dashDescription,
  };
  addDashboard({ dashName, dashDescription });
  dashboardsRef.push(dashboard);
};
