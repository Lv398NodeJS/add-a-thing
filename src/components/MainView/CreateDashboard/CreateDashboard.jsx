import React, { Component } from 'react';
import {
  Button, Container,
} from 'react-bootstrap';
import ModalToCreateDash from './ModalToCreateDash';
import '../../../App.scss';

export default class CreateDashboard extends Component {
  constructor() {
    super();

    this.state = {
      showComponent: false,
    };
  }

  handleShow = () => {
    this.setState({ showComponent: true });
  }

  render() {
    const { showComponent } = this.state;
    const { addDashboard } = this.props;
    return (
      <Container className="App">
        <Button className="createNewDash" variant="primary" onClick={this.handleShow}>
          Create new dashboard
          {showComponent ? <ModalToCreateDash className="modal-to-create" addDashboard={addDashboard} /> : null}
        </Button>
      </Container>
    );
  }
}
