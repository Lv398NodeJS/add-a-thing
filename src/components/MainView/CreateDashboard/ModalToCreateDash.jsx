import React, { Component } from 'react';
import { Modal, Form, Button } from 'react-bootstrap';
import db from '../../../fire';
import createNewDash from './PushDashboardtoDb';
import './ModalToCreateDash.scss';

class ModalToCreateDash extends Component {
  constructor() {
    super();
    this.state = {
      show: true,
      dashName: '',
      dashDescription: '',
    };
  }

  handleSaveButtonPush = () => {
    const { addDashboard } = this.props;
    const { dashName, dashDescription } = this.state;
    createNewDash(dashName, dashDescription, addDashboard);
    this.setState({
      show: false,
      dashName: '',
      dashDescription: '',
    });
  }

  handleClose = () => {
    this.setState({
      show: false,
    });
  }

  handleSave = (event) => {
    const { name, value } = event.target;
    this.setState({
      [name]: value,
    });
  }

  render() {
    const { show, dashName, dashDescription } = this.state;
    return (
      <Modal
        size="lg"
        show={show}
        onHide={this.handleClose}
        className="createBtnModal"
      >
        <Modal.Header closeButton>
          <Modal.Title>Create Dashboard</Modal.Title>
        </Modal.Header>
        <Form.Control
          className="modal-form"
          type="text"
          placeholder="Enter the name for the new dashboard"
          name="dashName"
          value={dashName}
          onChange={this.handleSave}
        />
        <Form.Group controlId="exampleForm.ControlTextarea1">
          <Form.Control
            className="modal-form"
            as="textarea"
            placeholder="Enter the description for the new dashboard"
            name="dashDescription"
            value={dashDescription}
            onChange={this.handleSave}
          />
        </Form.Group>
        <br />
        <Modal.Body>Click &apos;Save Changes&apos; to create a new dashboard</Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={this.handleClose}>
            Close
          </Button>
          <Button variant="primary" onClick={this.handleSaveButtonPush} disabled={!dashName}>
            Save Changes
          </Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

export default ModalToCreateDash;
