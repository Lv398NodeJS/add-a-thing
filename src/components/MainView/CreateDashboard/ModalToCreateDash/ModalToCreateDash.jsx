import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Modal, Form, Button } from 'react-bootstrap';
import createNewDash from './CreateNewDash';
import './ModalToCreateDash.scss';
import * as viewActions from '../../../../actions/mainViewActions';

export class ModalToCreateDash extends Component {
  constructor(props) {
    super(props);
    this.state = {
      show: props.show,
      dashName: '',
      dashDescription: '',
    };
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ show: nextProps.show });
  }

  handleClose = () => {
    const { closeModal } = this.props;
    closeModal();
    this.setState({
      show: false,
      dashName: '',
      dashDescription: '',
    });
  }

  handleSaveButtonPush = () => {
    const { closeModal, mainViewActions } = this.props;
    const { dashName, dashDescription } = this.state;
    createNewDash(dashName, dashDescription, mainViewActions.addDashboard);
    closeModal();
    this.setState({
      show: false,
      dashName: '',
      dashDescription: '',
    });
  }

  handleSave = (event) => {
    const { name, value } = event.target;
    this.setState({
      [name]: value,
    });
  }

  render() {
    const { dashName, dashDescription, show } = this.state;
    return (
      <Modal
        onHide={this.handleClose}
        show={show}
        size="lg"
        className="createBtnModal"
      >
        <Modal.Header closeButton>
          <Modal.Title>Create Dashboard</Modal.Title>
        </Modal.Header>
        <Form.Control
          className="modal-form dash-name"
          type="text"
          placeholder="Enter the name for the new dashboard"
          name="dashName"
          value={dashName}
          onChange={this.handleSave}
        />
        <Form.Group controlId="exampleForm.ControlTextarea1">
          <Form.Control
            className="modal-form dash-description"
            as="textarea"
            placeholder="Enter the description for the new dashboard"
            name="dashDescription"
            value={dashDescription}
            onChange={this.handleSave}
          />
        </Form.Group>
        <br />
        <Modal.Body>Click &apos;Save Changes&apos; to create a new dashboard</Modal.Body>
        <Modal.Footer>
          <Button
            className="close-button"
            variant="secondary"
            onClick={this.handleClose}
          >
            Close
          </Button>
          <Button
            className="save-changes"
            variant="primary"
            onClick={this.handleSaveButtonPush}
            disabled={!dashName}
          >
            Save Changes
          </Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  mainViewActions: bindActionCreators(viewActions, dispatch),
});

export default connect(
  null,
  mapDispatchToProps,
)(ModalToCreateDash);
