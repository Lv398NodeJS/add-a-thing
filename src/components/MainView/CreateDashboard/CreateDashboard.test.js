import React from 'react';
import { shallow, mount } from 'enzyme';
import CreateDashboard from './CreateDashboard';
import ModalToCreateDash from './ModalToCreateDash';

describe('CreateDashboard', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(<CreateDashboard debug />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly with no props', () => {
    const component = shallow(<CreateDashboard />);

    expect(component).toMatchSnapshot();
  });

  let mountedCreateBtn;

  const createDashboard = () => {
    if (!mountedCreateBtn) {
      mountedCreateBtn = mount(
        <CreateDashboard />,
      );
    }
    return mountedCreateBtn;
  };

  it('always renders the button', () => {
    const button = createDashboard().find('.createNewDash');
    expect(button.length).toBeGreaterThan(0);
  });

  it('opens modal window when clicked', () => {
    const wrapper = shallow(<CreateDashboard />);
    wrapper.find('.createNewDash').simulate('click');

    expect(wrapper.find(ModalToCreateDash).exists()).toEqual(true);
  });
});
