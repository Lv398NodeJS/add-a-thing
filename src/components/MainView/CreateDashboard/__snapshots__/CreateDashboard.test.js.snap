// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateDashboard should render correctly in "debug" mode 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateDashboard
    debug={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "as": "div",
      "children": <ForwardRef
        active={false}
        className="createNewDash"
        disabled={false}
        onClick={[Function]}
        type="button"
        variant="primary"
      >
        Create new dashboard
      </ForwardRef>,
      "className": "App",
      "fluid": false,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "active": false,
        "children": Array [
          "Create new dashboard",
          null,
        ],
        "className": "createNewDash",
        "disabled": false,
        "onClick": [Function],
        "type": "button",
        "variant": "primary",
      },
      "ref": null,
      "rendered": Array [
        "Create new dashboard",
        null,
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "active": false,
          "disabled": false,
          "type": "button",
          "variant": "primary",
        },
        "displayName": "Button",
        "render": [Function],
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "div",
        "fluid": false,
      },
      "displayName": "Container",
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "as": "div",
        "children": <ForwardRef
          active={false}
          className="createNewDash"
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
          Create new dashboard
        </ForwardRef>,
        "className": "App",
        "fluid": false,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": false,
          "children": Array [
            "Create new dashboard",
            null,
          ],
          "className": "createNewDash",
          "disabled": false,
          "onClick": [Function],
          "type": "button",
          "variant": "primary",
        },
        "ref": null,
        "rendered": Array [
          "Create new dashboard",
          null,
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "active": false,
            "disabled": false,
            "type": "button",
            "variant": "primary",
          },
          "displayName": "Button",
          "render": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "as": "div",
          "fluid": false,
        },
        "displayName": "Container",
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`CreateDashboard should render correctly with no props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateDashboard />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "as": "div",
      "children": <ForwardRef
        active={false}
        className="createNewDash"
        disabled={false}
        onClick={[Function]}
        type="button"
        variant="primary"
      >
        Create new dashboard
      </ForwardRef>,
      "className": "App",
      "fluid": false,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "active": false,
        "children": Array [
          "Create new dashboard",
          null,
        ],
        "className": "createNewDash",
        "disabled": false,
        "onClick": [Function],
        "type": "button",
        "variant": "primary",
      },
      "ref": null,
      "rendered": Array [
        "Create new dashboard",
        null,
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "active": false,
          "disabled": false,
          "type": "button",
          "variant": "primary",
        },
        "displayName": "Button",
        "render": [Function],
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "div",
        "fluid": false,
      },
      "displayName": "Container",
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "as": "div",
        "children": <ForwardRef
          active={false}
          className="createNewDash"
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
          Create new dashboard
        </ForwardRef>,
        "className": "App",
        "fluid": false,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": false,
          "children": Array [
            "Create new dashboard",
            null,
          ],
          "className": "createNewDash",
          "disabled": false,
          "onClick": [Function],
          "type": "button",
          "variant": "primary",
        },
        "ref": null,
        "rendered": Array [
          "Create new dashboard",
          null,
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "active": false,
            "disabled": false,
            "type": "button",
            "variant": "primary",
          },
          "displayName": "Button",
          "render": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "as": "div",
          "fluid": false,
        },
        "displayName": "Container",
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
